File: src/main/java/com/tang/intellij/lua/documentation/LuaDocumentationProvider.java
/*
 * Copyright (c) 2017. tangzx(love.tangzx@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tang.intellij.lua.documentation;

import com.intellij.lang.documentation.AbstractDocumentationProvider;
import com.intellij.lang.documentation.DocumentationProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNameIdentifierOwner;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import com.tang.intellij.lua.comment.LuaCommentUtil;
import com.tang.intellij.lua.comment.psi.*;
import com.tang.intellij.lua.comment.psi.api.LuaComment;
import com.tang.intellij.lua.psi.LuaCommentOwner;
import com.tang.intellij.lua.psi.LuaFuncBodyOwner;
import com.tang.intellij.lua.psi.LuaNameDef;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Documentation support
 * Created by tangzx on 2016/12/10.
 */
public class LuaDocumentationProvider extends AbstractDocumentationProvider implements DocumentationProvider {

    @Override
    public String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
        if (element instanceof LuaCommentOwner) {
            return genDocForCommentOwner((LuaCommentOwner) element);
        } else if (element instanceof LuaNameDef) {
            LuaCommentOwner owner = PsiTreeUtil.getParentOfType(element, LuaCommentOwner.class);
            if (owner != null)
                return genDocForCommentOwner(owner);
        }
        return super.generateDoc(element, originalElement);
    }

    private String genDocForCommentOwner(LuaCommentOwner owner) {
        StringBuilder sb = new StringBuilder();
        if (owner instanceof LuaFuncBodyOwner && owner instanceof PsiNameIdentifierOwner) {
            LuaFuncBodyOwner funcBodyOwner = (LuaFuncBodyOwner) owner;
            PsiNameIdentifierOwner methodDef = (PsiNameIdentifierOwner) owner;
            sb.append("<h1>");
            sb.append(methodDef.getName());
            sb.append(funcBodyOwner.getParamSignature());
        }

        LuaComment comment = LuaCommentUtil.findComment(owner);
        sb.append(genComment(comment));

        String doc = sb.toString();
        if (doc.length() > 0)
            return sb.toString();
        else
            return null;
    }

    private String genComment(LuaComment comment) {
        StringBuilder sb = new StringBuilder();
        if (comment != null) {
            PsiElement child = comment.getFirstChild();
            while (child != null) {
                IElementType elementType = child.getNode().getElementType();
                if (elementType == LuaDocTypes.STRING) {
                    sb.append(child.getText());
                    sb.append("<br>");
                } else if (elementType == LuaDocTypes.PARAM_DEF) {
                    LuaDocParamDef paramDef = (LuaDocParamDef) child;
                    LuaDocParamNameRef paramNameRef = paramDef.getParamNameRef();
                    if (paramNameRef != null) {
                        sb.append("<li><b>param</b> ");
                        sb.append(paramNameRef.getText());
                        sb.append(" ");
                        getTypeSet(paramDef.getTypeSet(), sb);
                        sb.append("<br>");
                    }
                } else if (elementType == LuaDocTypes.RETURN_DEF) {
                    LuaDocReturnDef returnDef = (LuaDocReturnDef) child;
                    LuaDocTypeList typeList = returnDef.getTypeList();
                    if (typeList != null) {
                        sb.append("<li><b>return</b> ");
                        List<LuaDocTypeSet> typeSetList = typeList.getTypeSetList();
                        for (LuaDocTypeSet typeSet : typeSetList) {
                            getTypeSet(typeSet, sb);
                            sb.append(" ");
                        }
                        sb.append("<br>");
                    }
                }
                child = child.getNextSibling();
            }
        }
        return sb.toString();
    }

    private void getTypeSet(LuaDocTypeSet typeSet, StringBuilder sb) {
        if (typeSet != null) {
            sb.append("(");
            List<LuaDocClassNameRef> nameRefList = typeSet.getClassNameRefList();
            String[] names = new String[nameRefList.size()];
            for (int i = 0; i < nameRefList.size(); i++) {
                names[i] = nameRefList.get(i).getText();
            }
            sb.append(String.join(", ", names));
            sb.append(")");
        }
    }
}

File: src/main/java/com/tang/intellij/lua/documentation/LuaDocumentationProvider.kt
/*
 * Copyright (c) 2017. tangzx(love.tangzx@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tang.intellij.lua.documentation

import com.intellij.codeInsight.documentation.DocumentationManagerUtil
import com.intellij.lang.documentation.AbstractDocumentationProvider
import com.intellij.lang.documentation.DocumentationProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiManager
import com.intellij.psi.util.PsiTreeUtil
import com.tang.intellij.lua.comment.psi.LuaDocTagClass
import com.tang.intellij.lua.comment.psi.LuaDocTagField
import com.tang.intellij.lua.editor.completion.LuaDocumentationLookupElement
import com.tang.intellij.lua.psi.*
import com.tang.intellij.lua.search.SearchContext
import com.tang.intellij.lua.stubs.index.LuaClassIndex
import com.tang.intellij.lua.ty.*

/**
 * Documentation support
 * Created by tangzx on 2016/12/10.
 */
class LuaDocumentationProvider : AbstractDocumentationProvider(), DocumentationProvider {

    private val renderer: ITyRenderer = object: TyRenderer() {
        override fun renderType(t: String): String {
            return if (t.isNotEmpty()) buildString { DocumentationManagerUtil.createHyperlink(this, t, t, true) } else t
        }
    }

    override fun getQuickNavigateInfo(element: PsiElement?, originalElement: PsiElement?): String? {
        if (element != null) {
            when (element) {
                is LuaTypeGuessable -> {
                    val ty = element.guessType(SearchContext(element.project))
                    return buildString {
                        renderTy(this, ty, renderer)
                    }
                }
            }
        }
        return super.getQuickNavigateInfo(element, originalElement)
    }

    override fun getDocumentationElementForLookupItem(psiManager: PsiManager, obj: Any, element: PsiElement?): PsiElement? {
        if (obj is LuaDocumentationLookupElement) {
            return obj.getDocumentationElement(SearchContext(psiManager.project))
        }
        return super.getDocumentationElementForLookupItem(psiManager, obj, element)
    }

    override fun getDocumentationElementForLink(psiManager: PsiManager, link: String, context: PsiElement?): PsiElement? {
        return LuaClassIndex.find(link, SearchContext(psiManager.project))
    }

    override fun generateDoc(element: PsiElement, originalElement: PsiElement?): String? {
        val sb = StringBuilder()
        val tyRenderer = renderer
        when (element) {
            is LuaParamNameDef -> renderParamNameDef(sb, element)
            is LuaDocTagClass -> renderClassDef(sb, element, tyRenderer)
            is LuaClassMember -> renderClassMember(sb, element)
            is LuaNameDef -> { //local xx

                renderDefinition(sb) {
                    sb.append("local <b>${element.name}</b>:")
                    val ty = element.guessType(SearchContext(element.project))
                    renderTy(sb, ty, tyRenderer)
                }

                val owner = PsiTreeUtil.getParentOfType(element, LuaCommentOwner::class.java)
                owner?.let { renderComment(sb, owner.comment, tyRenderer) }
            }
            is LuaLocalFuncDef -> {
                sb.wrapTag("pre") {
                    sb.append("local function <b>${element.name}</b>")
                    val type = element.guessType(SearchContext(element.project)) as ITyFunction
                    renderSignature(sb, type.mainSignature, tyRenderer)
                }
                renderComment(sb, element.comment, tyRenderer)
            }
        }
        if (sb.isNotEmpty()) return sb.toString()
        return super.generateDoc(element, originalElement)
    }

    private fun renderClassMember(sb: StringBuilder, classMember: LuaClassMember) {
        val context = SearchContext(classMember.project)
        val parentType = classMember.guessClassType(context)
        val ty = classMember.guessType(context)
        val tyRenderer = renderer

        renderDefinition(sb) {
            //base info
            if (parentType != null) {
                renderTy(sb, parentType, tyRenderer)
                with(sb) {
                    when (ty) {
                        is TyFunction -> {
                            append(if (ty.isColonCall) ":" else ".")
                            append(classMember.name)
                            renderSignature(sb, ty.mainSignature, tyRenderer)
                        }
                        else -> {
                            append(".${classMember.name}:")
                            renderTy(sb, ty, tyRenderer)
                        }
                    }
                }
            } else {
                //NameExpr
                if (classMember is LuaNameExpr) {
                    val nameExpr: LuaNameExpr = classMember
                    with(sb) {
                        append(nameExpr.name)
                        when (ty) {
                            is TyFunction -> renderSignature(sb, ty.mainSignature, tyRenderer)
                            else -> {
                                append(":")
                                renderTy(sb, ty, tyRenderer)
                            }
                        }
                    }

                    val stat = nameExpr.parent.parent // VAR_LIST ASSIGN_STAT
                    if (stat is LuaAssignStat) renderComment(sb, stat.comment, tyRenderer)
                }
            }
        }

        //comment content
        when (classMember) {
            is LuaCommentOwner -> renderComment(sb, classMember.comment, tyRenderer)
            is LuaDocTagField -> renderCommentString("  ", null, sb, classMember.commentString)
            is LuaIndexExpr -> {
                val p1 = classMember.parent
                val p2 = p1.parent
                if (p1 is LuaVarList && p2 is LuaAssignStat) {
                    renderComment(sb, p2.comment, tyRenderer)
                }
            }
        }
    }

    private fun renderParamNameDef(sb: StringBuilder, paramNameDef: LuaParamNameDef) {
        val owner = PsiTreeUtil.getParentOfType(paramNameDef, LuaCommentOwner::class.java)
        val docParamDef = owner?.comment?.getParamDef(paramNameDef.name)
        val tyRenderer = renderer
        if (docParamDef != null) {
            renderDocParam(sb, docParamDef, tyRenderer, true)
        } else {
            val ty = infer(paramNameDef, SearchContext(paramNameDef.project))
            sb.append("<b>param</b> <code>${paramNameDef.name}</code> : ")
            renderTy(sb, ty, tyRenderer)
        }
    }
}
